src/v2/mod.rs

```diff
@@ -2102,6 +2102,120 @@ impl Client {
+    ///
+    /// The return value is a pair of the absolute block height, and the hash,
+    /// of the block in which the instance was created. If the instance
+    /// cannot be found [`QueryError::NotFound`] is returned.
+    pub async fn find_instance_creation(
+        &mut self,
+        addr: ContractAddress,
+        start: Option<AbsoluteBlockHeight>,
+        end: Option<AbsoluteBlockHeight>,
+    ) -> QueryResult<(AbsoluteBlockHeight, BlockHash, InstanceInfo)> {
+        self.find_earliest_finalized(start, end, |mut client, height, bh| async move {
+            match client.get_instance_info(addr, &bh).await {
+                Ok(ii) => Ok(Some((height, bh, ii.response))),
+                Err(e) if e.is_not_found() => Ok(None),
+                Err(e) => Err(e),
+            }
+        })
+        .await
+    }
+
+    /// Find the first (i.e., earliers) finalized block whose slot time is no
+    /// later than the specified time. If a block is not found return
+    /// [`QueryError::NotFound`].
+    ///
+    /// If `start` and `end` are given the search is limited to that range,
+    /// inclusive. If the lower bound is not provided it defaults to 0, if the
+    /// upper bound is not provided it defaults to the last finalized block
+    /// at the time of the call.
+    pub async fn find_first_finalized_block_no_later_than(
+        &mut self,
+        time: chrono::DateTime<chrono::Utc>,
+        start: Option<AbsoluteBlockHeight>,
+        end: Option<AbsoluteBlockHeight>,
+    ) -> QueryResult<types::queries::BlockInfo> {
+        self.find_earliest_finalized(start, end, move |mut client, _, bh| async move {
+            let info = client.get_block_info(&bh).await?.response;
+            if info.block_slot_time >= time {
+                Ok(Some(info))
+            } else {
+                Ok(None)
+            }
+        })
+        .await
+    }
+
+    /// Find a block with lowest height that satisfies the given condition.
+    /// If a block is not found return [`QueryError::NotFound`].
+    ///
+    /// The `test` method should return `Some` if the object is found in the
+    /// block, and `None` otherwise. It can also signal errors which will
+    /// terminate search immediately.
+    ///
+    /// The precondition for this method is that the `test` method is monotone,
+    /// i.e., if block at height `h` satisfies the test then also a block at
+    /// height `h+1` does.
+    /// If this precondition does not hold then the return value from this
+    /// method is unspecified.
+    ///
+    /// If `start` and `end` are given the search is limited to that range,
+    /// inclusive. If the lower bound is not provided it defaults to 0, if the
+    /// upper bound is not provided it defaults to the last finalized block
+    /// at the time of the call.
+    pub async fn find_earliest_finalized<A, F: futures::Future<Output = QueryResult<Option<A>>>>(
+        &mut self,
+        start: Option<AbsoluteBlockHeight>,
```

Comment:
I don't like how this function reads when using optional values. A call like `client.find_earliest_finalized(None, None, ...)` is not readable IMO.

Consider making these mandatory and introduce helpers for getting the last finalized block height instead.
Maybe even introduce a block range type with functions for this. I imagine we will have more helpers working on block ranges.

Reply:
I disagree with that.

I have used these functions, and having mandatory arguments is just annoying because it leads to boilerplate on each call site where I don't care about the bounds (sometimes upper, sometimes lower).

But perhaps using `RangeBounds<AbsoluteBlockHeight>` trait would make you happier?

Reply:
Ok. I have used RangeBounds now. I think with the LSP support, at least the way I have it set up, the arguments to functions are displayed when calling the function. So there is nothing to remember. :)